add bisync crate for async support

omelia-iliffe opened the pull request:
A (large) PR to add Async support via the [bisync crate](https://crates.io/crates/bisync)

A number of public types have moved do to needing to seperate the method calls (that need to be either async or sync) and the types they return (as you don't need special AsyncPing, AsyncResponse, ..etc types ).

I also removed some top level exports to try to separate the client and device systems more.

The bisync pattern looks like this 
```
#[path = "."]
pub(crate) mod asynch {
	use crate::AsyncSerialPort as SerialPort;
	use bisync::asynchronous::*;
	mod bus;
	pub(crate) use bus::Bus;
}
#[path = "."]
pub(crate) mod sync {
	use crate::SerialPort;
	use bisync::synchronous::*;
	mod bus;
	pub(crate) use bus::Bus;
}
```

Instead of having one of these patterns at the top level I opted for multiple, one for Bus, Client, and Device. 
The modules asynch and sync are not exported, rather opting for structs with an `Async` prefix (ie `AsyncClient`)
This make it clearer imo.

The pattern also requires the use of `super` alot which I imagine might get tricky to maintain. Import the wrong thing without `super` and it might break in a hard to understand way.

Otherwise, it's great! I'm using it at the moment with embassy on the esp32 and enjoying it :)

omelia-iliffe committed:
add bisync crate for async support

omelia-iliffe committed:
add bisync crate for async support

omelia-iliffe committed:
removed bus_types, rename generic SerialPort to Port to avoid overlap

omelia-iliffe committed:
removed bus_types, rename generic SerialPort to Port to avoid overlap

omelia-iliffe committed:
added bisync to client

moved read_response_if_not_broadcast fn to Client

omelia-iliffe committed:
added bisync to client

moved read_response_if_not_broadcast fn to Client

omelia-iliffe committed:
moved instructions into client mod

moved packet_id and instruction_id in bus mod

omelia-iliffe committed:
moved instructions into client mod

moved packet_id and instruction_id in bus mod

omelia-iliffe committed:
removed common mod

changed visibility of some items

omelia-iliffe committed:
removed common mod

changed visibility of some items

omelia-iliffe committed:
fix docs

omelia-iliffe committed:
fix docs

omelia-iliffe committed:
fix tests and dynamixel2-cli

omelia-iliffe committed:
fix tests and dynamixel2-cli

omelia-iliffe committed:
move SerialPort to seperate module

omelia-iliffe committed:
move SerialPort to seperate module

omelia-iliffe committed:
fix features

omelia-iliffe committed:
fix features

omelia-iliffe committed:
restructure sync_read, bulk_read and scan, add async read_next/scan_next

de-vri-es commented on Cargo.toml:
```diff
@@ -25,17 +25,22 @@ alloc = []
 
 # Enable support for the `serial2` crate.
-serial2 = ["dep:serial2", "std"]
+serial2 = ["dep:serial2", "std", "dep:serial2-tokio", "tokio"]
```
Hmm, this will now pull in `tokio` even if you're writing a synchronous application.

Is there a way to avoid that? Could we simply move the `serial2-tokio` and `tokio` dependencies to the `serial2-tokio` feature?

omelia-iliffe committed:
re-export Client, AsyncClient, Device and AsyncDevice at crate root

Update the crate-level docs and README to link to the crate-root types.

omelia-iliffe committed:
decouple tokio from the synchronous serial2 feature

serial2 no longer enables serial2-tokio and tokio
add a serial2-tokio entry to the CI feature matrix

omelia-iliffe committed:
gate serial2-only tests and doctests behind the serial2 feature

gate the sync-API doctests, the bulk_write compile test and tests/common/real.rs
make integration-tests imply serial2
drop the now-unresolvable serial2::SerialPort link from the Client/Device docs
